src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +640,337 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
+		 *      approach (since we lose the opportunity to push-down the whole
+		 *      sub join tree into the workers) but is easier to implement.
+		 */
+
+		RecursivelyPlanDistributedJoinNode(((JoinExpr *) node)->larg,
+										   query, recursivePlanningContext);
+
+		RecursivelyPlanDistributedJoinNode(((JoinExpr *) node)->rarg,
+										   query, recursivePlanningContext);
+
+		return;
+	}
+
+	if (!IsA(node, RangeTblRef))
+	{
+		ereport(ERROR, (errmsg("unexpected join node type (%d)",
+							   nodeTag(node))));
+	}
+
+	RangeTblRef *rangeTableRef = (RangeTblRef *) node;
+	if (IsRTERefRecurring(rangeTableRef, query))
+	{
+		return;
+	}
+
+	RangeTblEntry *distributedRte = rt_fetch(rangeTableRef->rtindex,
+											 query->rtable);
+	if (distributedRte->rtekind == RTE_RELATION)
+	{
+		ereport(DEBUG1, (errmsg("recursively planning distributed relation %s "
+								"since it is part of a distributed join node "
+								"that is outer joined with a recurring rel",
+								GetRelationNameAndAliasName(distributedRte))));
+
+		PlannerRestrictionContext *restrictionContext =
+			GetPlannerRestrictionContext(recursivePlanningContext);
+		List *requiredAttributes =
+			RequiredAttrNumbersForRelation(distributedRte, restrictionContext);
+
+		ReplaceRTERelationWithRteSubquery(distributedRte, requiredAttributes,
+										  recursivePlanningContext);
+	}
+	else if (distributedRte->rtekind == RTE_SUBQUERY)
+	{
+		/*
+		 * We don't try logging the subquery here because RecursivelyPlanSubquery
+		 * will anyway do so if the query doesn't reference the outer query.
+		 */
+		ereport(DEBUG1, (errmsg("recursively planning the distributed subquery "
+								"since it is part of a distributed join node "
+								"that is outer joined with a recurring rel")));
+
+		/*
+		 * This was our only chance to plan given distributed subquery --known
+		 * to be part of the inner side of an outer join with a recurring rel--,
+		 * so we throw a hard error instead of a DeferredError if we can't
+		 * recursively plan given subquery.
+		 */
+		if (!RecursivelyPlanSubquery(distributedRte->subquery, recursivePlanningContext))
+		{
+			/*
+			 * RecursivelyPlanSubquery fails to plan a subquery only if it
+			 * contains references to the outer query. This means that, we can't
+			 * plan such outer joins (like <recurring LEFT OUTER distributed>)
+			 * if it's a LATERAL join where the distributed side is a subquery that
+			 * references the outer side, as in,
+			 *
+			 * SELECT * FROM reference
+			 * LEFT JOIN LATERAL
+			 * (SELECT * FROM distributed WHERE reference.b > distributed.b) q
+			 * USING (a);
+			 *
+			 * For this reason, we don't need to check such cases in
+			 * DeferredErrorIfUnsupportedRecurringTuplesJoin but instead
+			 * we capture such joins here and throw an error based on this
+			 * fact.
+			 */
+			Assert(ContainsReferencesToOuterQuery(distributedRte->subquery));
+
+			ereport(ERROR, (errmsg("cannot recursively plan the distributed "
```

Comment:
Hmm, I can see tests triggering this error. Do you know why code-cov is not showing?

Reply:
Not sure but from time-to-time codecov might fall behind branch's actual HEAD.

Reply:
Not a fan of the "cannot recursively plan" wording, that seems like an implementation detail.

Reply:
> Not a fan of the "cannot recursively plan" wording, that seems like an implementation detail that the user might not be familiar with.

What about ?
```c
			ereport(ERROR, (errmsg("cannot plan the query because the distributed "
								   "subquery involved in the inner side of the outer join "
								   "contains reference(s) to outer query")));
```


> (I'm also not quite sure about throwing this error in the recursive planner, but cannot quite come up with the reason not to)

Yeah, this sounded unusual to me too at the first glance but then the following convinced me for having this check here:
* As comment says, this was our only chance to plan given join.
* It's not only `RecursivePlanning..` function that returns / throws an error in the context of recursive planning but the check based on `NodeContainsSubqueryReferencingOuterQuery()` also does so